pytorch_lightning/core/lightning.py

```diff
@@ -1757,3 +1760,55 @@ def get_tqdm_dict(self) -> Dict[str, Union[int, str]]:
         rank_zero_warn("`get_tqdm_dict` was renamed to `get_progress_bar_dict` in v0.7.3"
                        " and this method will be removed in v1.0.0", DeprecationWarning)
         return self.get_progress_bar_dict()
+
+    def _auto_register_hparams(self):
+        """
+        Removes the need to pass in hparams. Instead, we register every argument in init
+        to the module with some caveats:
+        1. we don't overwrite the property if it already exists
+        2. we also store a module_arguments property for model loading and saving
+        """
+        # two frames back is the init of the child module
+        frame = inspect.currentframe()
+        args = frame.f_back.f_back.f_locals
```

Comment:
I mentioned this in https://github.com/PyTorchLightning/pytorch-lightning/issues/1735, but if my LightningModule was a subclass of another LightningModule, this wouldn't work right? We have to dynamically determine how many levels we should go since we always need to get to the leaf level.

Reply:
good point. any suggestions?
I guess we could always backtrack right up to the nn.Module?

Reply:
@yukw777 hadn't read that carefully actually haha. Good suggestion!

Reply:
(also, happy to co-author this since it's fairly involved)

Reply:
haha no worries. I'd be happy to help out.

I had to implement something similar at my job, and I ended up going with https://github.com/PyTorchLightning/pytorch-lightning/issues/1735#issuecomment-627746118. I tried to see if I can somehow automate the whole thing, but it was more trouble than it's worth, so I decided to keep things more "declarative". This does mean that a PL user would need to implement that abstract property, which makes `LightningModule` less transparent... We could make `LightningModule` a data class as @mateuszpieniak, but it's only available in 3.7, and it also makes `LightningModule` less transparent. It does seem like we do need to add something like this though, as it's impossible for PL to figure out whose `__init__()` args to save like this...

Reply:
how about i take a stab at a v1 and ping you to finish it haha.

Reply:
sounds good! looking forward to it!

Reply:
Re the dataclass, we could always depend on https://pypi.org/project/dataclasses/ to get dataclasses in 3.6.

I think adding a serialize_args abstract method makes sense. That's what I use at work (on a non lightning training pipeline) and it works pretty well. We have sensible serialization defaults, so it only needs to be overriden if the training module has custom, non serializable types.